### PR TITLE
[Draft] Adds a way to limits reference search by StructLiteral

### PR DESCRIPTION
**Pull request by mikhail-m1**
first draft for #2549

### CODE REVIEW

File: crates/ra_ide/src/references.rs [hunk 1]
```diff
@@ -35,15 +38,28 @@ pub use self::search_scope::SearchScope;
...
-    references: Vec<FileRange>,
+    declaration_kind: ReferenceKind,
+    references: Vec<Reference>,
+}
+
+#[derive(Debug, Clone)]
+pub struct Reference {
+    pub file_range: FileRange,
+    pub kind: ReferenceKind,
+}
```
Comment (kjeremy) [inline crates/ra_ide/src/references.rs]: I was just thinking we needed something like this for #2560

File: crates/ra_ide/src/references.rs [hunk 2]
```diff
@@ -58,24 +74,43 @@ impl ReferenceSearchResult {
...
 ) -> Option<RangeInfo<ReferenceSearchResult>> {
     let parse = db.parse(position.file_id);
     let syntax = parse.tree().syntax().clone();
+
+    let token = syntax.token_at_offset(position.offset);
+    let mut search_kind = ReferenceKind::Other;
+
+    if let TokenAtOffset::Between(ref left, ref right) = token {
+        if (right.kind() == SyntaxKind::L_CURLY || right.kind() == SyntaxKind::L_PAREN)
+            && left.kind() != SyntaxKind::IDENT
```
Comment (mikhail-m1) [inline crates/ra_ide/src/references.rs]: I added `left.kind() != SyntaxKind::IDENT` to ignore `fn_name()` case but now I think It should be something similar to look for `RecordLit` as I added below and add support for tuples.

**Comment by matklad** [comment]
This looks good to me!

I am slightly uneasy about the fact that `find_all_refs` is used the same way both for UI-facing `Analyzer::find_all_refs` as well as for an internal rename. I think we should maybe refactor that in the future (by making a more abstract reference search infra), but for the time being I think just not overengineering it is the right approach!

**Comment by kjeremy** [comment]
@matklad I use `find_all_refs` in #2698 too and was thinking something similar.

**Comment by matklad** [comment]
I think this is good to go actually! :)

bors r+

**Comment by matklad** [comment]
I've tested it and it works in Emacs, but not in VS Code: seems like vscode just does not trigger reference search if the cursor is on a punctuation :-(

**Comment by matklad** [comment]
Upstream vscode issue: https://github.com/microsoft/vscode-languageserver-node/issues/554

Not sure what we can do about this on our side :(

Gotta switch to Emacs I guess? :D

**Comment by matklad** [comment]
@mikhail-m1 I've rebased the branch from your repo onto master, to check if this makes bors behave

**Comment by matklad** [comment]
bors retry

**Pull request merged**